### ISSUE TITLE
Inconsistency in identifiers for apps: WebIDs and ClientIDs

### ISSUE THREAD

**Issue by tomhgmns** [solid/specification#463]
An app in Solid needs to have an identifier. The spec states that "WebIDs [...] are used as the primary identifier for users and applications." ([Solid Protocol, Section 9.1](https://solidproject.org/ED/protocol#webid)).

However, currently, apps are not only identified by their WebID, but also by their ClientID.

This leads to inconsistencies that hinder implementation.
* The Interop spec defines an endpoint for Agent Registration Discovery ([Solid Interop, Section 7.1.2](https://solidproject.org/ED/protocol#webid)) that should be able to lookup a registration for an agent based on the user's access token. This token contains the ClientID of the agent whilst the registries of the interop spec contain the WebID of the application
* There are two documents that describe an app (i.e. the WebID document and the ClientID document) and those documents contain duplicate information that is defined by different terms (e.g. interop:applicationName and oidc:clientName).

Considering that the Solid spec states that WebIDs "should be the primary identifier for users and applications", it seems logical to continue with the WebID of an app and drop the ClientID of an app.

One solution would be to merge the information in both documents into one document and serialise this information as desired using content negotiation.

However, in this solution, there would be a number of additional problems:
* Solid and in particular Solid OIDC and the WebID spec say nothing about accept headers
* This would still not solve the issue of different terms to describe the same information

Given the importance of keeping the specification consistent and the use of identifiers within Solid, I hope we can tackle this issue as fast as possible.

**Comment by elf-pavlik** [comment]
There is a related issue in Solid-OIDC https://github.com/solid/solid-oidc/issues/95

My preference would be to just provide `oidc` vocab and revert the specific JSON-LD requirement for ClientID Documents.

> There are two documents that describe an app (i.e. the WebID document and the ClientID document) and those documents contain duplicate information that is defined by different terms (e.g. interop:applicationName and oidc:clientName).

I don't think there was ever an intention to have two different documents. Solid-OIDC started with clients simply using WebIDs. Which later were replaced with ClientID and ClientID Document terminology.

Duplicate predicates in the `interop` namespace would be most likely removed. And we should pick one consistent set of predicated, be it from `oidc` namespace or whatever.

**Comment by justinwb** [comment]
Agree with picking one - no strong opinion on which - more important to just have one that's definitive.

**Comment by woutermont** [comment]
Agreed. The `client_id` and `webid` of an App should be the same URI, as was probably meant that way, and is semantically quite obvious. It then dereferences to a single document, which is both the WebID Profile and the Solid-OICD ClientID document. Together with aligning the vocabulary, this solves a number of problems and is semantically more elegant.

**Comment by woutermont** [comment]
On the most urgent matter of the WebID/ClientID Document format, we have a few options:
- Have the WebID spec change it to JSON-LD by default.
- Have the Solid-OIDC spec change theirs to Turtle by default. 
- Have the WebID spec mandate a `Accept: text/turtle` header. 
- Have the Solid-OIDC spec mandate a `Accept: application/ld+json` header. 

Personally, I don't have a strong preference. A consideration in favour of Turtle could be that it seems to be somewhat of a standard in many specs and existing applications. On the other hand, keeping JSON-LD as the default keeps up the hope of becoming compatible with OIDC (e.g. with it's Federation's Entity Statements).

**Comment by rubensworks** [comment]
> Solid and in particular Solid OIDC and the WebID spec say nothing about Accept headers

To keep specs orthogonal, I think this should remain this way. As long as they say that RDF MUST be returned, any standard RDF serialization should do.
This will also avoid the problem of every (Solid) spec having to repeat statements about RDF serializations, which may end up conflicting.

See also this long thread on mandatory versus recommended RDF serializations in the WebID CG: https://github.com/w3c/WebID/issues/3

**Comment by acoburn** [comment]
It is important not to conflate things.

Firstly, WebIDs can and are used to identify applications. Applications that use this mechanism will also need to be capable of managing private keys and issuing access tokens. In other words, these applications _do not use the OAuth2 `authorization_code` [flow](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.1)_. This is the case presently for many bots, scripts and other embedded devices that rely on flows other than `authorization_code`.

Second, for applications that _do_ use the `authorization_code` flow, the principal agent is identified with a WebID. This is typically an individual piloting an app. That agent is also delegating access to a particular app and making use of an authorization system that is defined outside of the Solid specification, namely: [OAuth2](https://datatracker.ietf.org/doc/html/rfc6749) and [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html). There is an important distinction between the agent identified with a WebID and the entity defined with a `client_id`. Calling both of those identifiers a WebID would create a different set of problems.

Either way, in order to work with OAuth2 and OpenID Connect, the [Solid-OIDC](https://solidproject.org/TR/oidc) specification uses a particular form of compact JSON-LD. This allows the Solid-OIDC specification to make use of and align with the [OIDC Dynamic Registration](https://openid.net/specs/openid-connect-registration-1_0.html) specification. The use of a JSON flavor for serializing this document is unlikely to change. The use of JSON-LD was a compromise in order to fit both the Solid _and_ the OpenID models, but leaving JSON altogether would break conformance with OpenID Connect: and adopting the WebID terminology for that does precisely this. There is [some discussion](https://github.com/solid/solid-oidc/issues/199) of changing the document format of Client Identifiers to align with the draft [OpenID Federation](https://openid.net/specs/openid-connect-federation-1_0.html#section-10.1) specification, but that would be JSON, not even JSON-LD and certainly not Turtle. Expecting the document format of OpenID application registrations to move away from JSON is mistaken, at least if we plan to continue to use OpenID Connect.

Because of the requirements of the WebID draft specification, it is not possible to call these Client Identifier URIs "WebIDs". And requiring these documents to be a non-JSON format conflicts with OpenID.

It is also important to note that there are many types of identifiers. WebIDs are one such type. Solid does, in fact, build significantly on  WebIDs, but at no point does the Solid Protocol specification state that WebIDs are the _only_ identifier that can be used for agents and applications. Some people in Solid are already using [DIDs](https://solid.github.io/did-method-solid/).

While I can appreciate that having multiple types of identifiers can introduce complexity, this is a _feature_ not a bug of decentralized ecosystems. It is also a feature of an ecosystem that looks to the future.

**Comment by elf-pavlik** [comment]
FYI in https://github.com/solid/solid-oidc/issues/95#issuecomment-1272326625 I'm bringing up the possibility to use `client_id` exclusively for OIDC and following mentioned by Aaron OpenID Federation spec. While having a distinct application website for a broader solid ecosystem. 

> That agent is also delegating access to a particular app and making use of an authorization system that is defined outside of the Solid specification, namely: [OAuth2](https://datatracker.ietf.org/doc/html/rfc6749) and [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html).

[Solid Application Interoperability](https://solid.github.io/data-interoperability-panel/specification/) introduces a base of an authorization system that allows end-user to authorize applications. 

> There is an important distinction between the agent identified with a WebID and the entity defined with a client_id. Calling both of those identifiers a WebID would create a different set of problems.

IMO we can denote both the end-user and the application (client) with WebID or DID or something else. We should be more careful not to use the term WebID without a context of what it denotes.

----

I think we need to organize all the requirements discussed here and look for proposals that take all of them into account. The mentioned separation between `client_id` and application webid could be one of the directions to explore. It might be useful to have a working document where we work on this problem. Conversations can become hard to work with once they grow longer.

**Comment by tomhgmns** [comment]
> There is an important distinction between the agent identified with a WebID and the entity defined with a `client_id`. Calling both of those identifiers a WebID would create a different set of problems.

I think I'm missing some background. It would help me if you could be a bit more precise on the following statements: 
- What is the distinction between the agent identified with a WebID and the entity defined with a `client_id` in case this agent is an application?
- Why is this distinction important?
- Could you give a few examples of the problems that would arise if an application's WebID and `client_id` would be the same?

**Comment by tomhgmns** [comment]
> To keep specs orthogonal, I think this should remain this way. As long as they say that RDF MUST be returned, any standard RDF serialization should do.
> This will also avoid the problem of every (Solid) spec having to repeat statements about RDF serializations, which may end up conflicting.

I agree with this point.

Be that as it may, let me rephrase my concern:
Solid and in particular Solid OIDC and the WebID spec say nothing about Accept headers _even though Solid-OIDC requires the use of JSON-LD and the WebID spec implies the use of Turtle_.

If both specs would allow any RDF serialisation, there wouldn't be a problem ;-)

**Comment by tomhgmns** [comment]
> While I can appreciate that having multiple types of identifiers can introduce complexity, this is a _feature_ not a bug of decentralized ecosystems. It is also a feature of an ecosystem that looks to the future.

This seems like a very strong statement without real references or examples. Could you give some examples of decentralised ecosystems in which a plethora of different identifiers have helped their growth?

Awaiting your response, I can give some counter examples of your statement:
- The email system only allows one type of identifier: an email address
- Identity systems based on blockchain only allow one type of identifier: a DID
- ...

> Some people in Solid are already using [DIDs](https://solid.github.io/did-method-solid/).
You've provided a link to a draft document about DIDs in Solid while I actually expected a link to a list of people that already used Solid with DIDs... To the best of my knowledge, no one actually uses Solid with DIDs in a production environment, but I hope that you can give me a counter example :-)

**Comment by rubensworks** [comment]
> Solid and in particular Solid OIDC and the WebID spec say nothing about Accept headers even though Solid-OIDC requires the use of JSON-LD and the WebID spec implies the use of Turtle.

> If both specs would allow any RDF serialisation, there wouldn't be a problem ;-)

That's a problem indeed (related to #454).

I just created #465 to tackle this specific problem.

**Comment by tomhgmns** [comment]
If we would decide against to use one type of identifier for an app, it would help if we could include a reference to the other in each of the documents.

By that I mean that:
- The Client ID document contains a reference to the WebID and
- The WebID document contains a reference to the Client ID

**Comment by woutermont** [comment]
@acoburn 

Thanks for the short recap for people who want to follow, but I don't see where we are conflating things.

> WebIDs can and are used to identify applications. [...] [T]hese applications do not use the OAuth2 `authorization_code` flow.

WebIDs can indeed be used for that purpose; we do so ourselves. However, I see no reason why WebID-identified applications could not use the `authorization_code` flow. In fact, if applications need to have an ID dereferenceable to a profile document (e.g. like in the Interoperability draft), a WebID seems like a good fit.

> There is an important distinction between the [principal] agent identified with a WebID and the entity defined with a `client_id`. Calling both of those identifiers a WebID would create a different set of problems.

You are right in pointing out the obvious distinction between the (social) agent using the client and the agent that is the client itself. However, I don't see why both could not be identified by (different) WebIDs. Again, the **client is an agent** that needs to be identified, and about which we want to declare some statements in a profile document.

> [Solid-OIDC] uses a particular form of compact JSON-LD [...] [to] align with the OIDC Dynamic Registration specification [and maybe the OpenID Federation specification]. The use of a JSON flavor for serializing this document is unlikely to change. [...] [L]eaving JSON altogether would break conformance with OpenID Connect: and adopting the WebID terminology for that does precisely this. [Because of this] it is not possible to call these Client Identifier URIs "WebIDs".

I'm not sure why you are holding that against us. **It is exactly this problem we want to raise** once more, as you did so yourself a year ago in https://github.com/w3c/WebID/issues/3 (before that issue sidetracked in a too general discussion about serialisations). 

If we presume that an application needs [A] an identifier that dereferences to a profile document containing Interop data, and [B] an identifier that dereferences to a document containing OIDC data, then why on earth would we have to create two different URI's pointing towards two different resources? (And likewise with any two other ID documents.) This holds even stronger when we realise that a lot of the information in both needs to be the same. Stubbornly keeping a specification that makes it more complex for developers is a bad practice, everybody knows that.

> It is also important to note that there are many types of identifiers [and] at no point does the Solid Protocol specification state that WebIDs are the only identifier that can be used for agents and applications. Some people in Solid are already using DIDs.

Definitely true, and rightly so. Now let's learn a bit about orthogonality from the [DID's take on representations](https://www.w3.org/TR/did-core/#representations), and kick those default obligations out of the WebID and DID:Solid specs! 

It isn't even hard; it's not like we're asking anyone to give up their preferred syntaxes. Just don't demand them on default requests:

> WebID requires that servers must at least be able to provide Turtle representation of profile documents, _when requested with `text/turtle` as the prefered format in the `Accept` header_.

For clarity we can then possibly add:

> WebID requires that servers must at least be able to provide a (i.e. any) W3C-recommended RDF-representation of profile documents, when requested without an `Accept` header or with an `Accept` header containing no acceptable formats.

**Comment by elf-pavlik** [comment]
> If we presume that an application needs [A] an identifier that dereferences to a profile document containing Interop data, and [B] an identifier that dereferences to a document containing OIDC data, then why on earth would we have to create two different URI's pointing towards two different resources? (And likewise with any two other ID documents.) This holds even stronger when we realize that a lot of the information in both needs to be the same.

In https://github.com/solid/solid-oidc/issues/199 and https://github.com/solid/solid-oidc/issues/95#issuecomment-1250087497 we discuss aligning the OIDC specific parts with  [example of Metadata with RP's Entity Configuration from *OpenID Connect Federation 1.0 - draft 22*](https://openid.net/specs/openid-connect-federation-1_0.html#section-4.1) 

If we do that, having two different IRI to identify the application in two different contexts might actually help us to meet different requirements of those two different contexts:

* `client_id` - specific to OIDC interactions following the OpenID Connect Federation 1.0 requirements
* `solid.appWebId` - used in the broad Solid context.

While we are at it we could also adjust claim in the ID Token to have matching `solid.userWebID`. As ecosystem evolve we could generalize further into `solid.appDID` and `solid.userDID`. (@woutermont I think your suggested something in those lines in https://github.com/solid/solid-oidc/issues/26)

**Comment by tomhgmns** [comment]
Just a kind reminder for the editors (i.e. @timbl @justinwb @dmitrizagidulin @kjetilk @csarven and @RubenVerborgh) that we keep running into this issue. 

It would be great if this issue could have some more visibility so that we can come to a decision.

**Comment by kjetilk** [comment]
Leading up to 0.9, we operated with milestones and project boards and a process to nominate and get issues on the board. Since I left Inrupt, I have not had the capacity to be a driving force for keeping the process going, but I believe we need to have that kind of structure and probably stick more closely to it too. At least, we can tag it.

**Comment by elf-pavlik** [comment]
@acoburn what do you think about my suggestion above https://github.com/solid/specification/issues/463#issuecomment-1279315991

**Comment by woutermont** [comment]
@elf-pavlik, personally I'm not a big fan of putting claims under a 'solid' object.

Also, if we use the federation entity statements, the clientid can just remain the app's WebID (which is what Tom is advocating for).

 I therefore think that this issue would be solved without further adjustments, when we make that step.

**Comment by woutermont** [comment]
Some considerations in light of the discussion during the CG Weekly.

---

### About the issue

There seems to be a huge amount of misunderstanding in this discussion, exemplified by the following statement.

> [@acoburn] [WebIDs and ClientIDs] are different semantically. [...] [A] WebID defines an agent (person), and then you go down the chain of delegation. [...] [A] client_id is identifying the client, not the user. If I am a user (person) I’d have a client which would be identified separately. 

This is true, but it is besides the question. This issue is NOT about the WebID of the person using an app versus the ClientID of the app itself. It is about the WebID _of the app_ versus the ClientID of the app. Both are identifiers of the very same entity (the app), so the question whether or not these can be the same IRI is completely justified.

The above is a theoretical, semantic question, of course, but @tomhgmns brings up a good practical consequence: if an app's WebID is different from its ClientID, a request containing a token with the WebID of the person using an app end the ClientID of the app itself (e.g. as in Solid-OIDC) cannot be used to retrieve data from the app's WebID Document (e.g. as in SAI). 

---

### About solution

We could indeed mandate mutual references in each document, as proposed by @elf-pavlik, but this seems just a hack when we're actually talking about the same entity. In the same line, we could put every piece of data from different specs in different resources, and let them all point to one another.

A more interesting option becomes available in the light of integration with OIDC Federation. @acoburn says the following:

> [With OIDC Federation] we can remove the whole client id from solid-oidc and use oidc federation. [...] If that happens that is pure json. _So I encourage to keep separate, conflating is trouble in long term._

While I gladly agree with the perspective this brings, I believe the conclusion here (my emphasis) is misguided. 

In the context of OIDC Federation Entity Discovery with Automatic Registration, which is what will subsume the current OIDC Dynamic Client Registration flow, the Client ID of an app will be its Entity Identifier (a globally unique URI). Such a URI has no specific restrictions except that the Entity Configuration (a self-issued Entity Statement) must be available at that URI's `.well-known/openid-federation` endpoint. 

We can thus perfectly envision an app with WebID / Entity Identifier / ClientID `https://example.app`, serving an RDF description on its base URL and a JSON Entity Configuration on `https://example.app/.well-known/openid-federation`.

**Comment by damooo** [comment]
Also there are semantic differences between client-id and webid of an app.

`WebId` being identifier for the agent, whether being as an app or person, that identity is unique `name` for the `agent` in entire world of discource on par with any other agent's name.

Where as `Client-id` is meaningful only for oidc clients,  and are `tied-against, and also often issued-by single oidc provider in it's own context`., And hence they are not the names an agent self-declare for themselves for representing them on open web. But are oidc-provider specific identifiers often issued and controlled by providers themselves, typically bound with assigned secrets for sole purpose of identification in their system. `Thus `client-id` for an app against one oidc-provider mostly will not be relevent against other provider, forget about openweb`.  They represent contextual contract between both provider and client(with their own TOC signed, possibly), where as web-id is self declared identifier.

**Comment by damooo** [comment]
Also one single app with name `Image transcriber app` can be registered with multiple oidc poviders. (Ascan be seen in current practice, sign-in-with google/github/twitter), each system assigning different client-id, secret pairs to same app identified by same literal name, and home page. They assign such an id after accepting their TOC. And can invalidate, if there is any misbehaviour.

`WebId` is just replacement for literal name of the app, that can be proven to be controlled by claimer.

Thus one app with a single literal-name/uri-name(web-id) can have multiple contracts with multiple providers, and thus multiple contractual identifiers. One-to-Many.

**Comment by acoburn** [comment]
@elf-pavlik I very much agree with https://github.com/solid/specification/issues/463#issuecomment-1279315991

I would also emphasize that, given that we are using the OpenID Connect framework for authentication, there are several distinct conceptual entities:

* Relying Party (RP) - OAuth 2.0 Client application requiring End-User Authentication and Claims from an OpenID Provider
* End-User - Human participant

In this context, the Relying Party is identified with a `client_id` and therefore is the Client Identifier. The End User is the entity with the `webid`. This entity may also be a bot -- i.e. a non-Human participant.

What is being discussed here is merging these two roles into a single role where the End User and the Relying Party are the same logical entity and are identified with a single URI.

It is possible to conflate these two roles into a single URI (and it is entirely possible to do this today in the context of the current Solid-OIDC specification), but this is not something I would recommend. These are distinct roles, and as such, I would highly recommend identifying those roles as independent entities.

**Comment by RubenVerborgh** [comment]
(I'm hesitant to reply since this is not my area of expertise, but I have been summoned above so I will attempt.)

Would all agree that the WebID and the ClientID are identifying different things? From @acoburn's answer, I understand that the WebID identifies the kind of software ("AppX", like an ISBN identifies a book), and the ClientID a specific instance ("version 3.7 of AppX running on machine ABC operated by user X, like a concrete copy of a book)?

If so, I see value in both; and we should in any case be extremely specific about what the WebID and ClientID identify.

I also think we've gotten confused as to what the exact question/problem is here. Original poster @tomhgmns asks about 1) why there are 2 identifiers (answer: because they are distinct things), and 2) their overlap (answer: we can't reinvent OIDC).

So could I ask @tomhgmns to be very specific about what "tackle this issue", and perhaps to do so in a different issue, as this one is getting very long and is already offers partial answers? I think "Agent Registration Discovery" is covered: pick the identifier you need (is it the app in general, or a specific instance); and what is the exact problem with the duplication, other than it is inconvenient but we can't circumvent OIDC? I.e., from my 200 feet view, I don't see what we're looking for here, and hence we're going in circles.

**Comment by elf-pavlik** [comment]
> What is being discussed here is merging these two roles into a single role where the End User and the Relying Party are the same logical entity and are identified with a single URI.

I think this is part of the misunderstanding. The discussion is only about URI denoting the application/client, it doesn't involve at all separate URI which denotes the End User.

**Comment by bblfish** [comment]
Perhaps a bit of logic can help.  A WebID identifies an Agent. It can identify a Person, in which case we expect the triple

```turtle
<#i> a foaf:Person .
```
It can identify a software instance in which case we should have a URL for the class of Apps
```turtle
<#myCal> a solid:App
```
where
```turtle
solid:App rdfs:subClassOf foaf:Agent
```
If one were to have an ISBN for an App then it should probably be a subclass of `solid:App` and perhaps,
since there can be many instances of it.

**Comment by acoburn** [comment]
To me, it seems that there are two reasonable solutions here, using the various specifications in their current form:

1. Use separate identifiers, one to align with WebID and the other to align with Solid-OIDC
2. Use the same identifier but make sure that the corresponding resource can be content-negotiated to produce both valid Turtle (to align with the draft WebID spec) and JSON-LD (to align with the Solid-OIDC spec)

**Comment by bblfish** [comment]
In today's meeting @woutermont made an interesting point that in the gov case they were working on they found the distinction between the Agent ID and the App Id to be unecessary. 

Whould that mean that they want something like this:

```turtle
<#x> a foaf:Person, solid:App .
```

Ie, and identifier that both refers to a Person and an App? That seems a bit weird. 
Perhaps I misunderstood.

**Comment by woutermont** [comment]
@RubenVerborgh, we're only going in circles because people get stuck on the same misinterpretation of this issue as @acoburn and @bblfish, which both I and @elf-pavlik have tried to [clear up](https://github.com/solid/specification/issues/463#issuecomment-1301069360): we do NOT want to say that a Person is an App. We are talking about assigning an App a WebID (cf. SAI) and a ClientID (cf. Solid-OIDC), and whether these two IDs can be the same URI.

The "issue" to me is semantic of nature, even though it has clearly presented practical effects: we cannot discover the one IRI via the other. "Tackling it" means: resolving the semantic friction and making that discovery possible.

To answer to @damooo, who is the only one who seems to have understood the core of the issue: a WebID and a ClientID (in the Solid-OIDC or OpenID Federation sense) are both globally unique identifiers. A WebID has semantics in the context of the WebID spec, a ClientID has semantics in the context of the (Solid-)OIDC (Federation) spec. Those respective semantics are:

- an HTTP URI which refers to an Agent and dereferences to a document describing the Agent;
- a URI that is globally unique and that is bound to one Entity.

These semantics are not incompatible; it is perfectly possible to have a 'ClientWebID': "a HTTP URI that is globally unique, that is bound to one Entity, which refers to an Agent and dereferences to a document describing the Agent." An Identifier adhering to that description would have meaning in both specs. 

Since in reality there is only a single Entity to which both the WebID and the ClientID correspond, it is also possible to assign this ClientWebID to it, and use it in both contexts. The only thing preventing us from doing that today is a bad interpretation of the WebID spec; but this will no longer be a problem once we incorporate OpenID Federation.

**Comment by bblfish** [comment]
> These semantics are not incompatible; it is perfectly possible to have a 'ClientWebID': "a HTTP URI that is globally unique, that is bound to one Entity, which refers to an Agent and dereferences to a document describing the Agent." An Identifier adhering to that description would have meaning in both specs.

Ok, I think I understood you @woutermont , I was just trying to make sure [above](https://github.com/solid/specification/issues/463#issuecomment-1301096543) to exclude the more bizzare interpretation. 

Perhaps can put it logically too. Let us say that we have a  `oidc:AppId` and a `:ClientWebID`. You are claiming these overlap or may even be identical concepts, thought to be distinct only for reasons  of syntax of the documents.  Ie that we can have

```Turtle
<#app> a :ClientWebID, oidc:AppId .
```

That actually seems quite reasonable to me, and indeed it was considered by @acoburn quite early on, since he was arguing to allow WebIDs to have turtle representations for just that reason  https://github.com/w3c/WebID/issues/3 .

**Comment by woutermont** [comment]
> indeed it was considered by @acoburn quite early on, since he was arguing to allow WebIDs to have turtle representations for just that reason https://github.com/w3c/WebID/issues/3 .

Exactly 😄 That's why I find it so bizar that he keeps involving the User in the discussion.

With regards to the logical notation, I don't think that you can put it that way, at least not in RDF: the identifier (`<#app>`) stands for the Entity it identifies, not for itself as an Identifier; so `<#app> a :ClientWebID .` actually claims that the App is a ClientWebID. You would need to be able to "parenthesize" the term `<#app>` to be able to talk about it on a meta-level: `Meta(<#app>) a :ClientWebID .`

**Comment by bblfish** [comment]
> With regards to the logical notation, I don't think that you can put it that way, at least not in RDF: the identifier (<#app>) stands for the Entity it identifies, not for itself as an Identifier; so <#app> a :ClientWebID . actually claims that the App is a ClientWebID. You would need to be able to "parenthesize" the term <#app> to be able to talk about it on a meta-level: Meta(<#app>) a :ClientWebID .

I agree. I just quickly grasped for those terms. But yes, something like `solid:App`  and `oidc:Client` would have been better terms  to use. Here `<#app>` is the the WebID - in relative form, as we assume that the context in which it is dereferenced will resolve it to a full absolute URL. It is a WebID because it identifies an agent, but we can be more precise about what subtype of agent it identifies. 

```turtle
<#app> a solid:App , oidc:Client .
```

Do we have official terms for any of those types yet?

**Comment by woutermont** [comment]
There is `interop:Application`, but it's not necessary to discuss the specific terms right now. An OpenID Federation "ClientID" (Entity Identifier) does not have to dereference to a profile document with a description of an Entity of a specific class.

**Comment by damooo** [comment]
In light of solid/solid-oidc#207,
To have same identifier as openweb identifier (web-id), and oidc-identiier, following impedance mismatch between their semantics must be resolved. OIDC-federation can indeed seems allow us to do that.

- **Control of the identifier**: Who mints, controls the identifier, and who can invalidate it. In classic OIDC, provider doesn't allow `bring your own client_id` model for identifier string. And entirely controls lifeycle of it, which is directly incompatible with webid semantics. OIDC-Federation entity identifiers seems allow the `bring-your-own` model. Thus following case can workout 
    1. Web id must deref to web-id doc in any serialization, to satisfy webid spec.
    2. OIDC-federation requirements based on id should be implemented. i.e to serve config at proper `.well-known` suffixed uri

- **Scope of the identifier**: WebId is globally unique for an app. Where as client-id need not be. As one app may want to be client of multiple providers, with distinct configuration for each. In above mapping, it is only possibble to have single oidc-federation config for an app if id is same as webid for every provider. Thus this limitation have to be worked around, or

**Comment by woutermont** [comment]
Thanks for the clear points, @damooo!

As you say, OpenID Federation allows a "bring-your-own" approach to ClientIDs (a.k.a. OIDC Dynamic Cient Registration, which is the current base of Solid-OIDC), and is thus indeed compatible with WebID (see also the examples at https://github.com/solid/solid-oidc/issues/207#issuecomment-1301735771 and [higher up this issue](https://github.com/solid/specification/issues/463#issuecomment-1300857441).

---

On the scope of the identifier, I must disagree with you. Both the WebID and the Entity Identifier (ClientID) must be globally unique (they are URIs), but neither must be unique for an app: a single app may have multiple WebIDs and/or Entity Identifiers (ClientID)s. 

For example: I can deploy an app on `https://my.app`, which sometimes identifies as `https://my.app/id1` and sometimes as`https://my.app/id2`, with the respective Entity Configurations `https://my.app/id1/.well-known/openid-federation` and `https://my.app/id2/.well-known/openid-federation`. Yes, this is a hack, but it should work fine.

**Comment by damooo** [comment]
Scope, in the sense not just of identifier string scope. But about scope of applicability, and satisfaction of criteria for entire setup ( deref + hosting config + ..).

One single setup as per above can seemlessly satisfy entire scope of webid. We can use that wherever webid is expected. But it is not the case for usage as oidc-entity uri. For that setup is bound to / scoped to one single federation with single config.

There can be many work-arounds as specified above. The ideal case of having unique webid, and offering services in many federations with custom configurations is not possible.

Ideally one should certainly be able to maintain single webid, to identify across whole spectrum of it's applicability. For users may want to say something about the app in federation agnostic way (review/ rate). It may be branding, accountability, etc...

**Comment by woutermont** [comment]
Correct, it is not "perfect" in that sense.

I do not think that this is behaviour which OpenID Federation wants to encourage. An Automatically Registered Entity has a fixed Entity Identifier. Doing another Automatic Registration at a second Federation using a different Entity Configuration will thus require a different Entity Identifier anyway; regardless of a combo with WebID. The problem you describe is thus a problem of OpenID Federation itself.

This issue only pointed out the fact that the combination of both roles by a single identifier was up till now not possible. With OpenID Federation it is; case closed. Let's leave it up to each implementer to decide if it is useful.

**Comment by tomhgmns** [comment]
Thank you all for your input.

As pointed out by @woutermont, OIDC federation (see https://github.com/solid/solid-oidc/issues/207), would indeed solve this issue. 

With OIDC federation, we could use the app's WebID as the app's ClientID and serve the contents that are currently in the ClientID document in a different document at another location.
